skorch/_doctor.py

```diff
@@ -0,0 +1,936 @@
+          For each top level module, all parameter names as a list of strings.
+
+        """
+        self.check_is_fitted()
+        names = {}
+        for module in self.module_names_:
+            if self.gradient_logs_[module]:
+                # using the reversed order because gradients are recorded from
+                # last to first, but first to last is more intuitive to show
+                # TODO: When dropping python 3.7, dict keys are reversible, so
+                # no need to call list(keys)
+                keys = list(self.gradient_logs_[module][0].keys())
+                names[module] = list(reversed(keys))
+            else:
+                names[module] = []
+        return names
+
+    def predict(self, X, **kwargs):
+        """Calls the ``predict`` method of the underlying net"""
+        return self.net.predict(X, **kwargs)
+
+    def predict_proba(self, X, **kwargs):
+        """Calls the ``predict_proba`` method of the underlying net"""
+        return self.net.predict_proba(X, **kwargs)
+
+    def score(self, X, y=None, **kwargs):
+        """Calls the ``score`` method of the underlying net"""
+        return self.net.score(X, y=y, **kwargs)
+
+    def _get_axes(self, axes=None, figsize=None, nrows=1, squeeze=False):
+        """Helper function to get empty axes for plotting"""
+        if axes is not None:
+            return axes
+
+        try:
+            import matplotlib.pyplot as plt
+        except ImportError as exc:
+            msg = (
+                "This feature requires matplotlib to be installed; "
+                "please install it first, e.g. using "
+                "'python -m pip install matplotlib'"
+            )
+            raise ImportError(msg) from exc
+
+        width = 12
+        height = 6
+
+        if figsize is None:
+            figsize = (width, nrows * height)
+
+        _, axes = plt.subplots(nrows, 1, figsize=figsize, squeeze=squeeze)
+        return axes
+
+    def plot_loss(self, ax=None, figsize=None, **kwargs):
+        """Plot the loss over each epoch.
+
+        Plots the training loss and, if present, the validation loss over time.
+
+        """
+        self.check_is_fitted()
+
+        ax = self._get_axes(ax, figsize=figsize, nrows=1, squeeze=True)
+        history = self.net.history
+        xvec = np.arange(len(history)) + 1
+        ax.plot(xvec, history[:, 'train_loss'], label='train', **kwargs)
+
+        if 'valid_loss' in history[0]:
+            ax.plot(xvec, history[:, 'valid_loss'], label='valid', **kwargs)
+
+        ax.set_xlabel('epoch')
+        ax.set_ylabel('loss')
+        ax.legend()
+        ax.set_title('loss over time')
+        return ax
+
+    def plot_activations(
+            self,
+            step=-1,
+            match_fn=None,
+            axes=None,
+            histtype='step',
+            lw=2,
+            bins=None,
+            density=True,
+            figsize=None,
+            **kwargs
```

Comment:
`kwargs` is routed directly to the `hist` plot. What do you think of the following API?

```python
def plot_activation(...,
    hist_kwargs=None,
):
    hist_kwargs = hist_kwargs or {}
```

This comment applies to the other `plot_*` methods.

Reply:
Not sure if I quite understand, you want to avoid having `**kwargs` at all but rather prefer to have `foo_kwargs` where `foo` corresponds to the kind of plot being generated? One reason why I chose to implement `kwargs` like this is because it is similar to the pandas plotting API, where you can also pass arbitrary `kwargs` and they are passed to the underlying plotting function.